src/ring_theory/polynomial/homogeneous.lean

```diff
@@ -225,6 +238,162 @@ noncomputable theory
 open_locale classical
 open finset
 
+open_locale direct_sum
+
+/-- A version of `homogeneous_component` that maps into `homogeneous_submodule σ R n`. -/
+def homogeneous_component' [comm_semiring R] (n : ℕ) :
+  mv_polynomial σ R →ₗ[R] homogeneous_submodule σ R n :=
+let f := finsupp.restrict_dom R R {d : σ →₀ ℕ | ∑ i in d.support, d i = n} in
+(submodule.of_le $ (homogeneous_submodule_eq_finsupp_supported _ _ _).symm.le).comp f
```

Comment:
I'm not sure I understand the question. Can you please elaborate?

Reply:
`homogeneous_component'` is the same map as `to_homogeneous_components`  immediately below, but the latter is more bundled.

However, the latter had to be built up from scratch, without using any of the existing finsupp machinery. This suggests either:
* There is some existing machinery I've missed
* There's a generalization here replacing `finsupp` with `add_monoid_algebra` that produces the stronger maps.

Reply:
My approach would be to define `to_homogeneous_components` first and then derive `homogeneous_component'` from that. My general approach to defining any morphisms is that you should almost never define a morphism by defining a function and then proving `map_mul` etc. You should pretty much always use things like `aeval`. I think `to_homogeneous_components` can be defined with `aeval`. I would probably also prove a different version of `aeval_monomial` that used the UMP of `finsupp` on the rhs instead of `finset.prod`. It's good practice to almost always use various hom-ext lemmas to prove equality of morphisms as well. Having said all that, I do agree with Johan that it would be better to define a characteristic predicate say that `mv_polynomial` is a direct_sum, than to define an isomorphism with a direct_sum, although you could of course use the isomorphism to define the direct_sum structure on polynomials.